fix(types): narrow down the return type of `defineConfig`

haoqunjiang opened the pull request:
### Description

Fixes compatibility with `mergeConfig`, as reported in https://github.com/vuejs/create-vue/issues/313

Before the change, `defineConfig` would return `UserConfigExport`, which is a union type. This would cause `mergeConfig` to fail, as it doesn't accept the `UserConfigFn` type in the union. (https://github.com/vitejs/vite/pull/13135)

After the change, `defineConfig` returns the same type that it receives, so it would exclude `UserConfigFn` from the return type whenever possible.

### Additional context

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [x] Bug fix
- [ ] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines) and follow the [PR Title Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Ideally, include relevant tests that fail without this PR but pass with it.

haoqunjiang committed:
fix(types): narrow down the return type of `defineConfig`

Fixes compatibility with `mergeConfig`, as reported in https://github.com/vuejs/create-vue/issues/313

Before the change, `defineConfig` would return `UserConfigExport`, which
is a union type. This would cause `mergeConfig` to fail, as it doesn't
accept the `UserConfigFn` type in the union. (https://github.com/vitejs/vite/pull/13135)

After the change, `defineConfig` returns the same type that it receives,
so it would exclude `UserConfigFn` from the return type whenever
possible.

patak-dev commented:
/ecosystem-ci run

haoqunjiang committed:
tests: remove unnecessary typecasts and `@ts-expect-error`s

patak-dev commented:
Only Qwik is failing compared to main. Iles and Histoire are known issues.

sodatea commented:
Update: I can confirm this is affected by the `["strictFunctionTypes"`](https://www.typescriptlang.org/tsconfig#strictFunctionTypes) configuration.

No tsconfig is configured for the `config.spec.ts` file, therefore it's not under strict mode, thus the local tests missed the bug.

bluwy commented:
FWIW this was also attempted at https://github.com/vitejs/vite/pull/12021 but reverted in https://github.com/vitejs/vite/pull/12077 also because of Qwik (and VitePress).

sodatea commented:
Further tests show that the `as any` casting in the Qwik config may be the culprit. https://github.com/BuilderIO/qwik/blob/190f0bcd7593745b58aef07054547551253d087b/packages/qwik-worker/vite.config.ts#L42
Removing that fixes the error.

I also managed to use the [`in out` operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-7.html#optional-variance-annotations-for-type-parameters) to fix the issue without modifying the downstream code, but ended up breaking some other use cases… But maybe it could work with more overloads. Let me try again.

haoqunjiang committed:
fix: revert to overload defineConfig

Based on #13792 but with slightly more verbose function style
configuration types.
We can make the generic version pass the tests with some hack, but it
is too permissive that it would allow users to pass in unknown fields.
The overload version is much more direct and simpler.

Thanks @xiaoxiangmoe for helping me debug this.

sodatea commented:
/ecosystem-ci run
